Fix summarise trips without trips

allenmichael099 opened the pull request:
@asiripanich Shankari and I noticed today that there was an error with some of the trip summary columns. All participants had the same value. Also, the participation period plot was uniform. After looking into it, I realized that the mistake was in how I implemented summarise_trips_without_trips. It was using the min and max dates for all trips, rather than the min and max for each user. 

It seems to be fixed but perhaps it should be more carefully tested. I threw together the beginnings of a test that for it in test-utils_tidy_data, and the columns are mostly the same. One version seems to round the timestamps to the nearest day while the other does not. n_active_days might appear in a different order.

Here are relevant screenshots of the issue  (see last trip date and number of days since app install):
![Wrong_data_upload](https://user-images.githubusercontent.com/56986329/126733885-5458ff75-be6a-44d8-ab0f-f0f4337d6be4.PNG)
![Wrong_particip_pd](https://user-images.githubusercontent.com/56986329/126733887-37542d1c-a6ea-4d60-a3e3-f9c05aa9e0f4.PNG)

And after the fix:
![Fixed_data_upload](https://user-images.githubusercontent.com/56986329/126733911-7925bb69-5aa2-4fe3-b0b9-daf829792964.PNG)
![Fixed_Particip_Pd](https://user-images.githubusercontent.com/56986329/126733915-d311400d-be7c-4084-87d7-a9abdde01c0d.PNG)

allenmichael099 committed:
Fix summarise_trips_without_trips and attempt tests

- trips summaries were not working properly, so I edited summarise_trips_without_trips
- added a testthat case to check whether the participants table is the same between trip summary methods

allenmichael099 committed:
Comment out summarise trips test

asiripanich commented:
@allenmichael099 just wondering why did you comment that test out? c821219

asiripanich committed:
Styler

allenmichael099 commented:
@asiripanich I figured it wasn't ready yet, and I didn't know what expectation to set.

asiripanich commented:
@allenmichael099 Could you please submit another PR with unit tests, after Aug is fine. Thanks!

asiripanich committed:
Merge branch 'master' into Fix_summarise_trips_without_trips

allenmichael099 committed:
Added mode_confirm column

- added mode_confirm to query_trip_dates
- added unconfirmed_summ line, as seen in Shankari's branch

allenmichael099 committed:
Made the unconfirmed/unlabeled column configurable

- added 'confirmed_user_input_column' to config, summarise_trips_without_trips, and query_trip_dates
- let's check that it makes sense to implement it this way

allenmichael099 committed:
Clear most summarise_trips_without_trips test cases

- still not getting the same first/last_trip_datetime, but only off by ~1 second

shankari commented on R/utils_tidy_data.R:
```diff
@@ -192,12 +198,22 @@ summarise_trips_without_trips <- function(participants, cons) {
 
   message("merging trip summaries with participants")
-  merge(participants, summ_trips, by = "user_id", all.x = TRUE)
+  merge(participants, summ_trips, by = "user_id", all.x = TRUE) %>% 
+    merge(., unconfirmed_summ, by="user_id", all.x = TRUE) %>%
+    .[is.na(unconfirmed), unconfirmed := 0]
```
Note to self: checked that this is the same as in my `master_specific_customization` branch

shankari commented on R/utils_tidy_data.R:
```diff
@@ -192,12 +198,22 @@ summarise_trips_without_trips <- function(participants, cons) {
-
+  
+  # Add the 'unconfirmed' column.
+  unconfirmed_summ <- trip_query %>%
+    .[is.na(trip_query[[confirmed_user_input_column]]), .(unconfirmed = .N), by = user_id]
+  
```
note to self, this makes sense and is also similar to my original code.

shankari commented on R/utils_tidy_data.R:
```diff
@@ -162,27 +162,33 @@ tidy_cleaned_trips_by_timestamp <- function(df) {
+  helper_trip_cols <- trip_query %>%
+    .[, .(
+      user_id = user_id,
+      start_fmt_time = lubridate::as_datetime(trip_query$data.start_fmt_time), # convert trip start datetimes to UTC
+      end_fmt_time = lubridate::as_datetime(trip_query$data.end_fmt_time) # convert trip end datetimes to UTC
```
note to self: this is similar to the original "before" code.

shankari commented on R/utils_query_database.R:
```diff
@@ -333,16 +333,19 @@ count_total_trips <- function(cons) {
-                                  "data.end_local_dt":true,
-                                  "data.end_fmt_time":true,
-                                  "user_id":true, "_id":false}'
+    fields = sprintf('{"data.start_local_dt":true,
+              "data.start_fmt_time":true,
+              "data.end_local_dt":true,
+              "data.end_fmt_time":true,
+              "%s":true,
+              "user_id":true, 
+              "_id":false}', confirmed_user_input_column)
```
note to self: these fields seem fine and the `sprintf` makes sense. Thanks R for using standard C function name.

shankari commented on R/utils_tidy_data.R:
```diff
@@ -162,27 +162,33 @@ tidy_cleaned_trips_by_timestamp <- function(df) {
+    # Now add columns generated from start/end_fmt_time
+    .[, ":="
+    (
+      start_local_time = purrr::map2(start_fmt_time, trip_query$data.start_local_dt.timezone, ~ format(.x, tz = .y, usetz = TRUE)) %>%
+        as.character(), # format returns a list, but lubridate::as_datetime needs it as a character string
+      end_local_time = purrr::map2(end_fmt_time, trip_query$data.end_local_dt.timezone, ~ format(.x, tz = .y, usetz = TRUE)) %>%
+        as.character()
```
Ditto

shankari reviewed (changes requested):
The code changes all make sense to me. Why is the test failing?

allenmichael099 committed:
Add a tolerance to the summarize trips with and without trips check.

asiripanich commented:
@allenmichael099  Now that we have `summarise_trips_without_trips()`, do we still need `summarise_trips()`

asiripanich committed:
- Upgrade to testthat 3 edition.

asiripanich committed:
Datetime columns should be POSIXct.

asiripanich committed:
Snap test

asiripanich committed:
Merge branch 'master' into Fix_summarise_trips_without_trips

allenmichael099 commented:
@asiripanich I kept summarise_trips to test whether the output of summarise_trips_without_trips is different than it should be. It might still be useful to have summarise_trips to compare with the output from summarise_trips_without_trips after any future changes to summarise_trips_without_trips.

asiripanich commented:
> 
> 
> @asiripanich I kept summarise_trips to test whether the output of summarise_trips_without_trips is different than it should be. It might still be useful to have summarise_trips to compare with the output from summarise_trips_without_trips after any future changes to summarise_trips_without_trips.

That is what I thought you were trying to do. Instead of that, we can use `testthat` to take a snapshot of the output and test against that. See https://github.com/asiripanich/emdash/blob/master/tests/testthat/_snaps/utils_tidy_data.md.
